### ISSUE TITLE
Stringent table detection

### ISSUE THREAD

**Issue by cinjon** [tabulapdf/tabula-java#30]
In the pdf at this link - https://www.dropbox.com/s/shrl08zbkwxzcni/17.pdf?dl=0 - the auto table detect works except for the first row of the first table. It seems to miss that one completely. However, if I expand the border a little bit on each side, then it detects it. 

Is there a way to have the autodetect be a little less stringent?

**Comment by jazzido** [comment]
Not possible right now, without manually specifiying the area of interest. 

A possible patch, off the top of my head, could be scaling the detected area by a small percentage to make sure that it includes the entire contents. 

What do you think?

**Comment by cinjon** [comment]
I'm unsure what you mean. Are you suggesting scaling it proportionally so that a 100x150 rectangle becomes a 120x180 rectangle with 10/15 padding on each side (for example)?

If so, yep, I think that would work. What's your plan if it doesn't include all the contents? And how would you detect that?

**Comment by jazzido** [comment]
> If so, yep, I think that would work. What's your plan if it doesn't include all the contents? And how would you detect that?

If i'm not mistaken, this issue happens because the detected area is *exactly* the same as the bounding ruling lines. Growing the area of interest by a tiny amount, making sure that it both includes every ruling line inside it, and does not intersect with elements *outside* might work just fine.

**Comment by jeremybmerrill** [comment]
This is related to https://github.com/tabulapdf/tabula-extractor/issues/99. (just xrefing)

**Comment by cinjon** [comment]
Looks like the xref was from late May. Is there another issue holding up a fix?

**Comment by jeremybmerrill** [comment]
I think the holdup is the question of which solution to implement (and day jobs, that's a holdup too... :-/ ):

1. treating the selection boundary as a table boundary
2. expanding the selection by a couple percentage points

**Comment by cinjon** [comment]
What do you mean by "treating the selection boundary as a table boundary"? Is that what @jazzido was saying wouldn't work in the xref?

**Comment by jeremybmerrill** [comment]
Yes, @jazzido thinks it wouldn't work, but I think it would... :)

I'm reluctant to try buffering selections by an arbitrary percentage or pixel count, just because it seems hacky and inelegant.

My current suspicion is that this is just a complex off-by-one error where the *bottom* of some lines (maybe thick lines with a >=2px stroke) is returned out of the autodetect algorithm, so the line itself is not included in the selection when the coordinates are passed to the Lattice extraction method. Confirming this suspicion: in @cinjon's PDF, you only need to slightly expand the top and left lines to get the top row to be properly included (so there doesn't appear to be a problem with the bottom/right portions of the autodetected selection).

**Comment by jeremybmerrill** [comment]
It looks like this issue doesn't occur in Tabula stable/1.0.x branch (i.e. the current release branch). Are you all seeing that as well?

For what it's worth, the autodetected selection in stable is  `311.65082454582216, 67.40983668899537, 502.9741600408173, 512.5902048835754` (TLBR)

**Comment by cinjon** [comment]
How did you see that? I just tried Tabula 1.0.1 and 1.0.0 from the releases
page <https://github.com/tabulapdf/tabula/releases> and neither method
(Stream or Lattice) in the UI worked right. The Stream one doesn't map to
the correct places and the Lattice one doesn't find all the cells.

**Comment by jeremybmerrill** [comment]
Hmm, that is puzzling. I don't think there's a difference between the head of stable/1.0.x, but now I'm not sure, because I'm seeing this problem in Tabula 1.0.1...

**Comment by jazzido** [comment]
> Yes, @jazzido thinks it wouldn't work, but I think it would... :)

It might work, let's test it. Although its elegance, compared to the buffering solution, it's certainly debatable :)

**Comment by jeremybmerrill** [comment]
I'll grant you that they're both inelegant...

I do wonder if we can get something quite elegant by examining how we deal with thick-stroked lines... (or by examining how PDFBox tests for line L's inclusion in rectangle R when the line L overlaps an edge of R...)

**Comment by jazzido** [comment]
> I do wonder if we can get something quite elegant by examining how we deal with thick-stroked lines... (or by examining how PDFBox tests for line L's inclusion in rectangle R when the line L overlaps an edge of R...)

AFAIK, PDFBox does not know anything about that. There are primitives in PDF for drawing both lines and rectangles, and PDFBox just let us interpret them. One of the problems is that some PDFs draw lines as thin rectangles, and some others use proper lines.

**Comment by jeremybmerrill** [comment]
Ah, well, our code then, might be too stringent in testing for inclusion?
